Blog: Hands on with Portals (#862)
GoogleChrome/web.dev: src/site/content/en/blog/hands-on-portals/index.md

---

> title: "Hands-on with Portals: seamless navigations on the Web"

Author: Thank you for the comment! Added!

author: uskay
hero: hero.png
alt: A logo of Portals

> description: |

Author: Thank you for the comment!! First of all, I want to check if the suggestion you are making (`author` to `authors`) is correct. It fails the build.
```bash
Problem writing Eleventy templates: (more in DEBUG output)
> Having trouble rendering njk template ./src/site/content/en/blog/index.njk (TemplateContentRenderError)
> (unknown path)
  Error:
      Can't create Author component without an author argument.
      author was null or undefined.
     (Template render error):
    Template render error: (unknown path)
```

> description: |

Author: Swapping back to `author` for now.

> description: |

Reviewer: ah sorry, yes `authors` is correct but that change was made last night so you may have forked before it went in.

> description: |

Reviewer: You can leave it as author and I can fix it before merging so you don't have to mess with rebasing.

  An article on explaining how to use the new HTML element: Portals
tags:
  - portals
  - UX
---

We’ve long stressed the importance of web performance. While it is tempting to focus on individual pages, it is also critical to maintain a good user experience as your users navigate across your pages. A new upcoming web platform API called [Portals](https://github.com/WICG/portals) can help you with that.

> ![A gif of a portals demo](hero.gif)

Author: Got that!! I'll add the class to the image!

# What Portals enables
Single Page Applications with an [App Shell model](https://developers.google.com/web/fundamentals/architecture/app-shell) are one way of achieving seamless navigations. Having a static shell on the page, like the header, footer and side nav, and dynamically changing the content with a transition provides continuity to the user experience in ways that a classic navigation experience in Multiple Page Applications can’t. In particular, the navigation experience on Multiple Page Applications is plagued by frequent flashes of blank or unfinished paints. While SPA result in a better user experience, there are many reasons why you might not want to choose SPA as your frontend architecture. Maybe you don’t have the necessary expertise in your team, or you might just want to simplify the front-end and strategically select the more scalable and productivity friendly MPA architecture. The Web is open to any types of websites and each one of them should have a good way to satisfy the users with liquid smooth experience.

> Portals is a new HTML element that can help you keep your front-end simple while allowing seamless navigations with custom transitions. The best part is that Portals is quite easy to work with. Think of it as an iframe in that it allows for embedding, but unlike an iframe it also comes with a feature to directly navigate into its content. Seeing is believing: please first check out what we showcased in Chrome Dev Summit last year.

Author: I think we are using the word "Portals" as a proper noun. e.g. 
> Portals is a very early proposal
https://github.com/WICG/portals/blob/master/README.md

> Portals is a new HTML element that can help you keep your front-end simple while allowing seamless navigations with custom transitions. The best part is that Portals is quite easy to work with. Think of it as an iframe in that it allows for embedding, but unlike an iframe it also comes with a feature to directly navigate into its content. Seeing is believing: please first check out what we showcased in Chrome Dev Summit last year.

Reviewer: I'm wondering if the usage in your GitHub example is because it's referring to the portals proposal (singular), and not portals themselves. As someone unfamiliar with the background and reading the article for the first time it looked like a grammatical error.

If it's referring to a new HTML element it feels a bit odd to refer to it as  a proper noun. Like saying "Buttons is" or "Iframes is". In the youtube video one of the speakers says, "So, what are portals? Portals are like iframes."

> Portals is a new HTML element that can help you keep your front-end simple while allowing seamless navigations with custom transitions. The best part is that Portals is quite easy to work with. Think of it as an iframe in that it allows for embedding, but unlike an iframe it also comes with a feature to directly navigate into its content. Seeing is believing: please first check out what we showcased in Chrome Dev Summit last year.

Author: Makes sense. I'll fix it.

> Portals is a new HTML element that can help you keep your front-end simple while allowing seamless navigations with custom transitions. The best part is that Portals is quite easy to work with. Think of it as an iframe in that it allows for embedding, but unlike an iframe it also comes with a feature to directly navigate into its content. Seeing is believing: please first check out what we showcased in Chrome Dev Summit last year.

Author: I changed this as well but now I'm thinking I'm wrong

> Whereas the improved experience with Portals **were** pre-rendering the next chapter

In this case, the subject would be experience and I should have kept it to **was**?

<div style="width:100%; padding-top: 56.25%; position: relative;">
  <iframe style="width:100%; height: 100%;position: absolute; top: 50%; left: 50%; transform: translate(-50%,-50%);" src="https://www.youtube.com/embed/Ai4aZ9Jbsys?start=1081" frameborder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
</div>

> As you saw in the manga viewer demo, the current experience was based on classic navigations where the reader had to wait with a blank screen until the browser finished to render the next chapter. Whereas the improved experience with Portals was pre-rendering the next chapter, triggered an animation on user interaction, and finally activated the portal to create a magical seamless experience. The demo didn’t show it (because we used Chrome DevTools mobile simulator 😛) but when the pre-rendered chapter on the bottom left corner fully animated and covered the screen, the URL also switched to the next chapter.

Author: Let me leave the "current" so that it has a contrast between before (current version) and after (Portals version). I agree with other suggestions so fixing it (removing "finally" and "magical").

Before Portals, we could have rendered another page using an iframe. We could also have added animations to move the frame around the page. But iframes won’t let you navigate into their content. Portals close this gap, enabling interesting use cases.

# Try out Portals in Chrome Canary

> You can try out Portals in Chrome Canary by flipping an experimental flag (chrome://flags/#enable-portals). Once Portals is enabled, confirm in DevTools that you have the new shiny `HTMLPortalElement`.

Author: Adding `chrome://flags/#embed-portals` as text & link.

![A screenshot of the DevTools console showing the HTMLPortalElement](HTMLPortalElement.png)

Let’s walk through a basic example.

```javascript
// Create a portal with the wikipedia page, and embed it
// (like an iframe). You can also use <portal> tag instead.
portal = document.createElement('portal');
portal.src = 'https://en.wikipedia.org/wiki/World_Wide_Web';
portal.style = '...';
document.body.appendChild(portal);

// When the user touches the preview (embedded portal):
// do fancy animation, e.g. expand …
// and finish by doing the actual transition
portal.activate();
```

> It’s that simple. Try this code in the DevTools console, the wikipedia page should open up. If you wanted to build something like we showed in CDS, the following snippet will be of interest.

Reviewer: ```suggestion
It’s that simple. Try this code in the DevTools console, the wikipedia page should open up. If you wanted to build something like we showed at Chrome Dev Summit, the following snippet will be of interest.
```

![A gif of preview portal style demo](portal-preview-demo.gif)

> ```javascript

Reviewer: Instead of this snippet, would you be ok turning this into a glitch using https://glitch.com/ ?

Then we can embed a running example on the site. You could also include a running version of the previous wikipedia example.

Author: If so, can I 
1. Keep this snippet as is (as this is just for reference code)
2. Remove the GIF below and add the running glitch (https://uskay-portals-demo.glitch.me)

Author: Rob, I was considering how I should be using glitch embed but I think I'll stick with not using glitch embeds this time. The reason is: Portals can only be used in Chrome Canary now with an experimental flag on. I suppose most of the people will be reading this on unsupported browsers which doesn't really make sense to have the demos/snippets embedded as glitch (it will not run or show fallback stuff which is not ideal).

    left: calc(20px + 50% * ${initialScale} - 50%);
    z-index: 10000;
  }
  .portal-reveal {
    transform: scale(1.0);
    bottom: 0px;
    left: 0px;
  }
  .fade-in {
    opacity: 1.0;
  }
`;
const portal = document.createElement('portal');
// Let’s navigate into the WICG Portals spec page
portal.src = 'https://wicg.github.io/portals/';
portal.addEventListener('click', evt => {
  // Animate the portal once user interacts
  portal.classList.add('portal-reveal');
});
portal.addEventListener('transitionend', evt => {
  if (evt.propertyName == 'bottom') {
    // Activate the portal once the transition has completed
    portal.activate();
  }
});
document.body.append(style, portal);

// Waiting for the page to load.
// using setTimeout is a suboptimal way and it’s best to fade-in
// when receiving a load complete message from the portal via postMessage
setTimeout(_ => portal.classList.add('fade-in'), 2000);
```

It is also easy to do feature detection to progressively enhance a website using Portals.

```javascript
if ('HTMLPortalElement' in window) {
  // If this is a platform that has Portals...
  const portal = document.createElement('portal');
  ...
}
```

> If you want to quickly experience what Portals feels like, try using [uskay-portals-demo.glitch.me](https://uskay-portals-demo.glitch.me). 1) Enter a URL you want to preview, 2) the page will then be embedded as a portal element, 3) click on the preview 4) The preview will be activated after an animation.

Reviewer: ```suggestion
If you want to quickly experience what Portals feel like, try using [uskay-portals-demo.glitch.me](https://uskay-portals-demo.glitch.me).

1. Enter a URL you want to preview.
1. The page will then be embedded as a portal element.
1. Click on the preview.
1. The preview will be activated after an animation.
```

Reviewer: You can use `1.` over and over and markdown will figure out the right number to put next to the ordered list item.

Author: Changed it to ordered list!

Author: BTW, super neat feature of keeping it all "1" but having the right numbers.

> ![A gif of using the glitch demo](glitch.gif)

Reviewer: Suggest wrapping this with the `w-screenshot` class.

Author: I will do that!

> # Check out the spec

Reviewer: ```suggestion
## Check out the spec
```

> We are actively discussing the spec in [WICG](https://github.com/WICG/portals/blob/master/explainer.md). You can check out the details [here](https://wicg.github.io/portals/) but overall there are three important features that builds up the element.

Reviewer: ```suggestion
We are actively discussing [the spec](https://wicg.github.io/portals/) in the Web Incubation Community Group (WICG). To quickly get up to speed, take a look at [the explainer](https://github.com/WICG/portals/blob/master/explainer.md). These are the three important features to familiarize yourself with:
```

>  - [The portal element](https://wicg.github.io/portals/#the-portal-element) … The HTML element itself. The API is very simple. It consists of the `src` attribute, the `activate` function and an interface for messaging (`postMessage`). activate has an optional argument to pass data to the portal upon activation.

Reviewer: ```suggestion
 - [The portal element](https://wicg.github.io/portals/#the-portal-element) … The HTML element itself. The API is very simple. It consists of the `src` attribute, the `activate` function and an interface for messaging (`postMessage`). `activate` takes an optional argument to pass data to the portal upon activation.
```

>  - [The portal host interface](https://wicg.github.io/portals/#the-portalhost-interface) … This adds a `portalHost` object to the window object. This lets you check if the page is embedded as a portal element. It also provides an interface for messaging (`postMessage`) back to the host.

Reviewer: ```suggestion
 - [The portal host interface](https://wicg.github.io/portals/#the-portalhost-interface) … Adds a `portalHost` object to the `window` object. This lets you check if the page is embedded as a portal element. It also provides an interface for messaging (`postMessage`) back to the host.
```

>  - [The PortalActivateEvent interface](https://wicg.github.io/portals/#the-portalactivateevent-interface) … An event being fired when the portal is activated. There is a neat function called `adoptPredecessor` which you can use to retrieve the previous page as a `<portal>  ` element. This allows you to create seamless navigations and composed experiences between two pages.

Reviewer: ```suggestion
 - [The PortalActivateEvent interface](https://wicg.github.io/portals/#the-portalactivateevent-interface) … An event that fires when the portal is activated. There is a neat function called `adoptPredecessor` which you can use to retrieve the previous page as a `<portal>` element. This allows you to create seamless navigations and composed experiences between two pages.
```

> Let’s look beyond the basic usage pattern. Here is a non-exhaustive list of you can achieve with Portals along with sample code.

Reviewer: ```suggestion
Let’s look beyond the basic usage pattern. Here is a non-exhaustive list of what you can achieve with Portals along with sample code.
```

> ## Customize the style when embedded as a portal element

Reviewer: If you want you could turn some or all of these into embedded glitches to let folks preview the code snippet. That's totally optional. Only if you think it will help the article. I realize at the end there's a video which kind of ties it all together but I think having things folks can play around with can be fun.

Author: Thanks for the suggestion! Unfortunately, these are not so useful to just have them individually running on glitch. I would rather want to keep the snippet as is. And lead them to the WICG demos repo in the end (and the vid).

Reviewer: sounds good.

});
```
## Knowing your page was adopted as a predecessor
```javascript
// The activate function returns a Promise.
// When the promise resolves, it means that the portal has been activated.
// If this document was adopted by it, then window.portalHost will exist.
portal.activate().then(_ => {
  // Check if this document was adopted into a portal element.
  if (window.portalHost) {
    // You can start communicating with the portal element
    // i.e. listen to messages
    window.portalHost.addEventListener('message', evt => {
      // handle the event
    });
  }
});
```
By combining all the features supported by Portals, you can build really fancy user experiences. For instance the demo below demonstrates how Portals can enable a seamless user experience between a website and a third party embed content. 

<div style="width:100%; padding-top: 56.25%; position: relative;">
  <iframe style="width:100%; height: 100%;position: absolute; top: 50%; left: 50%; transform: translate(-50%,-50%);" src="https://www.youtube.com/embed/4JkipxFVE9k" frameborder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
</div>

> [Here is the link to the GitHub repo](https://github.com/WICG/portals/tree/master/demos/portal-embed-demo) and you can play around with it. Be creative and have fun!!

Reviewer: ```suggestion
[Here is the link to the GitHub repo](https://github.com/WICG/portals/tree/master/demos/portal-embed-demo) which you can play around with. Be creative and have fun!!
```

> # Use cases and the plan

Reviewer: ```suggestion
## Use cases and the plan
```

> I hope you liked this brief tour of Portals! We can’t wait to see what you can come up with. For instance, you might want to start using Portals for non-trivial navigations such as: prerendering the page for your best-seller product from a product category listing page. Another important thing to know is that Portals can be used in cross-origin navigations, just like iframes. So, if you have multiple websites that cross reference one another, you can also use Portals to create seamless navigations between two different websites. This cross-origin use case is very unique to Portals, and can even improve the user experience of SPA.

Reviewer: ```suggestion
I hope you liked this brief tour of Portals! We can’t wait to see what you can come up with. For instance, you might want to start using Portals for non-trivial navigations such as: prerendering the page for your best-seller product from a product category listing page. Another important thing to know is that Portals can be used in cross-origin navigations, just like iframes. So, if you have multiple websites that cross reference one another, you can also use Portals to create seamless navigations between two different websites. This cross-origin use case is very unique to Portals, and can even improve the user experience of SPAs.
```

> Note that Portals is still in early stages so not everything is working yet (that’s why it’s behind an experimental flag). That said, it’s ready for early experimentation in Chrome Canary. Feedback from the community is crucial to the design of new APIs, so please try it out and tell us what you think! You can check the current limitations [here](https://bugs.chromium.org/p/chromium/issues/detail?id=957836) and if you have any feature requests, or feedback, please head over to the [WICG GitHub repo](https://github.com/WICG/portals/issues).

Reviewer: ```suggestion
Note that Portals are still in early stages so not everything is working yet (that’s why it’s behind an experimental flag). That said, it’s ready for early experimentation in Chrome Canary. Feedback from the community is crucial to the design of new APIs, so please try it out and tell us what you think! You can check the current limitations on [the Chromium bug tracker](https://bugs.chromium.org/p/chromium/issues/detail?id=957836) and if you have any feature requests, or feedback, please head over to the [WICG GitHub repo](https://github.com/WICG/portals/issues).
```

Reviewer: Suggest not using `here` as link text for accessibility reasons.
https://web.dev/labels-and-text-alternatives/#links

Author: Understood! I haven't really thought of a11y so thanks for pointing out!!

---
title: "Hands-on with Portals: seamless navigations on the Web"
subhead: |

>   A new web platform API called Portals can help you build good navigation UX across your site.

Author: Thank you! Is this OR or AND? I love your idea but should it be:
- Build good navigation UX with the Portals API.
OR
- Learn how to build good navigation UX with the Portals API.

Or are you saying it should be "Build good navigation UX with the Portals API. Learn how to build good navigation UX with the Portals API."

I'd love to change this to your suggestion.

>   A new web platform API called Portals can help you build good navigation UX across your site.

Reviewer 2: Or-- not and. Just wanted to give you some choices :).

>   A new web platform API called Portals can help you build good navigation UX across your site.

Author: Thank you! I'd take the "Learn how to build good navigation UX with the Portals API."!!

author: uskay
hero: hero.png
alt: A logo of Portals
description: |

>   Portals are a new HTML element that can help you keep your front-end simple while allowing seamless navigations with custom transitions. In this article we will walk through how to creatively use Portals to improve the user experience on the Web.

Author: SGTM! Thank you so much for helping this out!

tags:
  - post
  - portals
  - ux
---

> We’ve long stressed the importance of web performance. While it is tempting to focus on individual pages, it is also critical to maintain a good user experience as your users navigate across your site. A new web platform API called [Portals](https://github.com/WICG/portals) can help you with that.

Author: SGTM!

<img class="w-screenshot" src="hero.gif" alt="Showing a demo of seamless navigations between two sites using Portals">

## What Portals enables

> Single Page Applications (SPAs) with an [App Shell model](https://developers.google.com/web/fundamentals/architecture/app-shell) are one way of achieving seamless navigations. Having a static shell on the page, like the header, footer and side nav, and dynamically changing the content with a transition provides continuity to the user experience in ways that a classic navigation experience in Multiple Page Applications (MPAs) can’t. In particular, the navigation experience on Multiple Page Applications is plagued by frequent flashes of blank or unfinished paints. While SPAs result in a better user experience, there are many reasons why you might not want to choose SPA as your front-end architecture. Maybe you don’t have the necessary expertise in your team, or you might just want to simplify the front-end and strategically select the more scalable and productivity friendly MPA architecture. The Web is open to any type of website and each one of them should have a good way to satisfy the users with liquid smooth experience.

Author: Thank you! Can I add something like this?

Portals is a new HTML element that can help you keep your front-end simple while allowing seamless navigations with custom transitions. Like dynamic navigations in Single Page Applications with an App Shell model,
Portals let you dynamically change content with transitions but **you can keep your Multiple Page Application architecture**.
But the best part about Portals is they are quite easy to work with.

Basically I 100% support your suggestion. I just want to have a good way to contrast SPA and MPA and tell the audience that you can still do cool navigations with MPA. Can I have your advice here?

Also, Rob mentioned that might be good to treat Portals as plural. I see you are using it as singular. What is the recommendation here?
https://github.com/GoogleChrome/web.dev/pull/862/files/087f0437099528b828d18ad11687b44b1c01790d#r280608890

> Single Page Applications (SPAs) with an [App Shell model](https://developers.google.com/web/fundamentals/architecture/app-shell) are one way of achieving seamless navigations. Having a static shell on the page, like the header, footer and side nav, and dynamically changing the content with a transition provides continuity to the user experience in ways that a classic navigation experience in Multiple Page Applications (MPAs) can’t. In particular, the navigation experience on Multiple Page Applications is plagued by frequent flashes of blank or unfinished paints. While SPAs result in a better user experience, there are many reasons why you might not want to choose SPA as your front-end architecture. Maybe you don’t have the necessary expertise in your team, or you might just want to simplify the front-end and strategically select the more scalable and productivity friendly MPA architecture. The Web is open to any type of website and each one of them should have a good way to satisfy the users with liquid smooth experience.

Reviewer 2: Yeah, I like your rewrite. It's cool to treat Portals as plural, but then the verbs should be plural too. So Portal is a new HTML element that can help...
I think singular is good for describing the element. You could even tag it.... `<portal>` is a new...

And then when you are referring to them in use, you'd use plural...

Portals let you...

Basically when you introduce the element aspect of it, it'd be weird, i think to say Portals are new html elements... and yet, Portals is new... verb usage is off.

Make sense?

> Single Page Applications (SPAs) with an [App Shell model](https://developers.google.com/web/fundamentals/architecture/app-shell) are one way of achieving seamless navigations. Having a static shell on the page, like the header, footer and side nav, and dynamically changing the content with a transition provides continuity to the user experience in ways that a classic navigation experience in Multiple Page Applications (MPAs) can’t. In particular, the navigation experience on Multiple Page Applications is plagued by frequent flashes of blank or unfinished paints. While SPAs result in a better user experience, there are many reasons why you might not want to choose SPA as your front-end architecture. Maybe you don’t have the necessary expertise in your team, or you might just want to simplify the front-end and strategically select the more scalable and productivity friendly MPA architecture. The Web is open to any type of website and each one of them should have a good way to satisfy the users with liquid smooth experience.

Author: Yes, so I was actually pointing out to the verbs. So it okay (feel natural for English speakers) to keep it as "Portals is a new HTML element..." ?

**Portals** are a new HTML element that can help you keep your front-end simple while allowing seamless navigations with custom transitions. The best part is that Portals are quite easy to work with. Think of them like an iframe in that they allow for embedding, but unlike an iframe they also come with features to directly navigate to their content. Seeing is believing: please first check out what we showcased at Chrome Dev Summit last year.

<div style="width:100%; padding-top: 56.25%; position: relative;">
  <iframe style="width:100%; height: 100%;position: absolute; top: 50%; left: 50%; transform: translate(-50%,-50%);" src="https://www.youtube.com/embed/Ai4aZ9Jbsys?start=1081" frameborder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
</div>

> As you saw in the manga viewer demo, the current experience was based on classic navigations where the reader had to wait with a blank screen until the browser finished to render the next chapter. Whereas the improved experience with Portals were pre-rendering the next chapter, triggered an animation on user interaction, and activated the portal to create a seamless experience. The demo didn’t show it (because we used Chrome DevTools mobile simulator 😛) but when the pre-rendered chapter on the bottom left corner fully animated and covered the screen, the URL also switched to the next chapter.

Author: SGTM!

Before Portals, we could have rendered another page using an iframe. We could also have added animations to move the frame around the page. But iframes won’t let you navigate into their content. Portals close this gap, enabling interesting use cases.

## Try out Portals in Chrome Canary

> You can try out Portals in Chrome Canary by flipping an experimental flag: [`chrome://flags/#enable-portals`](chrome://flags/#enable-portals). Once Portals are enabled, confirm in DevTools that you have the new shiny `HTMLPortalElement`.

Author: Thank you!

<img class="w-screenshot" src="HTMLPortalElement.png" alt="A screenshot of the DevTools console showing the HTMLPortalElement">

Let’s walk through a basic example.

> 

Author: This, yes, I understand that it's good to have that structure you suggested but given the interest of time, can I please keep it as is? If needed, I could work on it after I/O (need to fly in a bit). Also the interesting thing is that, you can copy paste this whole snippet and paste in the console and it works. I kind of want to keep the unity there.

> 

Author: I totally understand your suggestion (thank you so much!) and would want to work on it but (very minor thing but) I'm keeping the style sync with https://github.com/WICG/portals/blob/master/demos/portal-embed-demo/README.md which we just landed and I'm not sure if I can find time to replace all the style within the timeframe of I/O. Will it be okay to keep it as is for now?

```javascript
// Create a portal with the wikipedia page, and embed it
// (like an iframe). You can also use <portal> tag instead.
portal = document.createElement('portal');
portal.src = 'https://en.wikipedia.org/wiki/World_Wide_Web';
portal.style = '...';
document.body.appendChild(portal);

// When the user touches the preview (embedded portal):
// do fancy animation, e.g. expand …
// and finish by doing the actual transition
portal.activate();
```

> It’s that simple. Try this code in the DevTools console, the wikipedia page should open up. If you wanted to build something like we showed at Chrome Dev Summit, the following snippet will be of interest.

Reviewer 2: Suggestion to bring the sentence about 'If you wanted to build' to come after the image showing the wikipedia page. So that the sentence is coming just before the code.. Also, as previously mentioned, not all users will watch the video from Chrome Dev Summit. If possible, can you introduce the snippet, maybe saying something like:

"The following snippet shows you how to use portals in a way that's similar to the portals demo at the start of this article:"

Reviewer 2: Also, is there a way to break up this code a little bit? You could introduce the parts of it in stages, and then show it all together at the end? A glitch tutorial would really work well here. But I know time is precious!

Author: First bit, will work on it! 

Second bit, this, yes, I understand that it's good to have that structure you suggested but given the interest of time, can I please keep it as is? If needed, I could work on it after I/O (need to fly in a bit). Also the interesting thing is that, you can copy paste this whole snippet and paste in the console and it works. I kind of want to keep the unity there.

Author: Regarding your first comment, can I change it as

"If you wanted to build something like we showed at Chrome Dev Summit which works just like the demo above, 
the following snippet will be of interest."

And having it after the GIF? 1) this snippet doesn't represent the very first GIF in the article 2) I agree with making it understandable for the people that didn't play the YT video so I added "which works just like the demo above" to also reference the GIF above.

    left: calc(20px + 50% * ${initialScale} - 50%);
    z-index: 10000;
  }
  .portal-reveal {
    transform: scale(1.0);
    bottom: 0px;
    left: 0px;
  }
  .fade-in {
    opacity: 1.0;
  }
`;
const portal = document.createElement('portal');
// Let’s navigate into the WICG Portals spec page
portal.src = 'https://wicg.github.io/portals/';
portal.addEventListener('click', evt => {
  // Animate the portal once user interacts
  portal.classList.add('portal-reveal');
});
portal.addEventListener('transitionend', evt => {
  if (evt.propertyName == 'bottom') {
    // Activate the portal once the transition has completed
    portal.activate();
  }
});
document.body.append(style, portal);

// Waiting for the page to load.
// using setTimeout is a suboptimal way and it’s best to fade-in
// when receiving a load complete message from the portal via postMessage
setTimeout(_ => portal.classList.add('fade-in'), 2000);
```

It is also easy to do feature detection to progressively enhance a website using Portals.

```javascript
if ('HTMLPortalElement' in window) {
  // If this is a platform that have Portals...
  const portal = document.createElement('portal');
  ...
}
```

> If you want to quickly experience what Portals feel like, try using [uskay-portals-demo.glitch.me](https://uskay-portals-demo.glitch.me). Be sure you access with Chrome Canary and turn on the experimental flag!

Reviewer 2: +1 to Rob's suggestion to make this a proper glitch embed. It would be a very cool thing for developers to be able to play with the code a bit in the glitch.

Author: As I commented in Robs feedback, I also understand having glitch embed is nice, but at the same time this feature requires Chrome Canary with an experimental flag on which means most of the readers will see an app that doesn't work (although I have a fallback message). So I wast thinking of keeping this as is for now. WDYT?

Reviewer: If you're in a rush I think it's ok to land it as-is. But I do think adding the Glitch would be pretty cool. In the above example you're linking to a Glitch and telling the user to turn the feature flag on. What you could do in a follow up revision is embed the Glitch, and use the feature detection steps to detect if the user has Portals enabled, and if not, show them a link to the feature flag.

Likely this post will be up for a long time so when Chrome ships Portals it'll be cool to have a running example in the page that people can edit.

1. Enter a URL you want to preview.
1. The page will then be embedded as a portal element.
1. Click on the preview.
1. The preview will be activated after an animation.

<img class="w-screenshot" src="glitch.gif" alt="A gif of using the glitch demo of using Portals">

## Check out the spec
We are actively discussing [the spec](https://wicg.github.io/portals/) in the Web Incubation Community Group (WICG). To quickly get up to speed, take a look at [the explainer](https://github.com/WICG/portals/blob/master/explainer.md). These are the three important features to familiarize yourself with:

>  - [The portal element](https://wicg.github.io/portals/#the-portal-element) … The HTML element itself. The API is very simple. It consists of the `src` attribute, the `activate` function and an interface for messaging (`postMessage`). `activate` takes an optional argument to pass data to the portal upon activation.

Reviewer 2: Small nit to switch up the "..." with a colon inside the markdown link. And to do this for each bullet.

Author: I will do that! Thank you!

 - [The portal host interface](https://wicg.github.io/portals/#the-portalhost-interface) … Adds a `portalHost` object to the `window` object. This lets you check if the page is embedded as a portal element. It also provides an interface for messaging (`postMessage`) back to the host.
 - [The PortalActivateEvent interface](https://wicg.github.io/portals/#the-portalactivateevent-interface) … An event that fires when the portal is activated. There is a neat function called `adoptPredecessor` which you can use to retrieve the previous page as a `<portal>` element. This allows you to create seamless navigations and composed experiences between two pages.

Let’s look beyond the basic usage pattern. Here is a non-exhaustive list of what you can achieve with Portals along with sample code.

> ### Customize the style when embedded as a portal element

Reviewer 2: I really like these little snippets.

// You can optionally add data to the argument of the activate function
portal.activate({data: {'somekey': 'somevalue'}});

// The portal content will receive the portalactivate event
// when the activate happens
window.addEventListener('portalactivate', evt => {
  // Data available as evt.data
  const data = evt.data;   
});
```
### Retrieving the predecessor
```javascript
// Listen to the portalactivate event
window.addEventListener('portalactivate', evt => {
  // ... and creatively use the predecessor
  const portal = evt.adoptPredecessor();
  document.querySelector('someElm').appendChild(portal);
});
```
### Knowing your page was adopted as a predecessor
```javascript
// The activate function returns a Promise.
// When the promise resolves, it means that the portal has been activated.
// If this document was adopted by it, then window.portalHost will exist.
portal.activate().then(_ => {
  // Check if this document was adopted into a portal element.
  if (window.portalHost) {
    // You can start communicating with the portal element
    // i.e. listen to messages
    window.portalHost.addEventListener('message', evt => {
      // handle the event
    });
  }
});
```

> By combining all the features supported by Portals, you can build really fancy user experiences. For instance the demo below demonstrates how Portals can enable a seamless user experience between a website and a third party embed content. 

Reviewer 2: Small nit, missing comma after 'For instance'.

Author: Thank you!

<div style="width:100%; padding-top: 56.25%; position: relative;">
  <iframe style="width:100%; height: 100%;position: absolute; top: 50%; left: 50%; transform: translate(-50%,-50%);" src="https://www.youtube.com/embed/4JkipxFVE9k" frameborder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
</div>

[Here is the link to the GitHub repo](https://github.com/WICG/portals/tree/master/demos/portal-embed-demo) which you can play around with. Be creative and have fun!!

## Use cases and the plan

> I hope you liked this brief tour of Portals! We can’t wait to see what you can come up with. For instance, you might want to start using Portals for non-trivial navigations such as: pre-rendering the page for your best-seller product from a product category listing page. Another important thing to know is that Portals can be used in cross-origin navigations, just like iframes. So, if you have multiple websites that cross reference one another, you can also use Portals to create seamless navigations between two different websites. This cross-origin use case is very unique to Portals, and can even improve the user experience of SPAs.

Reviewer 2: Suggestion to remove 'I'. So, 'Hope you liked this brief tour of Portals!'.

Author: I will do that.

> I hope you liked this brief tour of Portals! We can’t wait to see what you can come up with. For instance, you might want to start using Portals for non-trivial navigations such as: pre-rendering the page for your best-seller product from a product category listing page. Another important thing to know is that Portals can be used in cross-origin navigations, just like iframes. So, if you have multiple websites that cross reference one another, you can also use Portals to create seamless navigations between two different websites. This cross-origin use case is very unique to Portals, and can even improve the user experience of SPAs.

Reviewer 2: Start new paragraph with "Another important thing to know".

Author: Thank you! Will fix!!

> Note that Portals are still in early stages so not everything is working yet (that’s why it’s behind an experimental flag). That said, it’s ready for early experimentation in Chrome Canary. Feedback from the community is crucial to the design of new APIs, so please try it out and tell us what you think! You can check the current limitations on [the Chromium bug tracker](https://bugs.chromium.org/p/chromium/issues/detail?id=957836) and if you have any feature requests, or feedback, please head over to the [WICG GitHub repo](https://github.com/WICG/portals/issues).

Reviewer 2: Maybe put this note in it's own section called "Feedback welcome". Then remove 'Note that' from start of paragraph within the section. 'Portals are still in early stages ...

Author: Agreed. Will do that!
